{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import meerkat as mk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "> src@0.0.1 dev\n",
      "> vite dev \"--port\" \"7861\"\n",
      "\n",
      "\n",
      "  VITE v3.0.3  ready in 411 ms\n",
      "\n",
      "  ➜  Local:   http://localhost:7861/\n",
      "  ➜  Network: use --host to expose\n"
     ]
    }
   ],
   "source": [
    "network, register_api = mk.interactive_mode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1\"\n",
       "            height=\"1\"\n",
       "            src=\"http://localhost:7861/network/register?api=http://127.0.0.1:7860\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x1049cddc0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "register_api()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/eyubogln/code/meerkat/meerkat/columns/file_column.py:221: FutureWarning: Reading a LambdaColumn stored in a format that will not be supported in the future. Please re-write the column to the new format.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from meerkat.interactive.app.src.lib.interfaces.match_gallery import MatchGalleryInterface\n",
    "\n",
    "\n",
    "dp = mk.DataPanel.read(\"data/imagenette_clustered.mk\")\n",
    "dp = dp[\"img\", \"path\", \"label\", \"split\", \"img_path\", \"label_idx\", \"clip(img)\"]\n",
    "interface = MatchGalleryInterface(\n",
    "    dp,\n",
    "    id_column=\"img_path\",\n",
    "    against=\"img\",\n",
    "    main_column=\"img\",\n",
    "    tag_columns=[\"label\", \"split\"],\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"1000\"\n",
       "            src=\"http://localhost:7861/interface?id=c0011f49c7934b6ea4bf0d5d5fea7cd7\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x14fc88160>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:    Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 50, in schema\n",
      "    id=pivot.obj.id, columns=_get_column_infos(dp, columns), nrows=len(dp)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Triggering store 9fc0fe9580ff4b71b9dc807c93e049db label\n",
      "Triggering store 96afb86ae8b541bda4e6bad23fa385ea []\n",
      "Triggering store 06275167d86740a180e9dcb44ee249b6 label\n",
      "Triggering store e33ce80002744000b4364479eab1dcb2 []\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2022-09-01 13:42:57,579][ERROR][uvicorn.error:406] :: Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 50, in schema\n",
      "    id=pivot.obj.id, columns=_get_column_infos(dp, columns), nrows=len(dp)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n",
      "ERROR:    Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 50, in schema\n",
      "    id=pivot.obj.id, columns=_get_column_infos(dp, columns), nrows=len(dp)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n",
      "[2022-09-01 13:42:57,608][ERROR][uvicorn.error:406] :: Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 50, in schema\n",
      "    id=pivot.obj.id, columns=_get_column_infos(dp, columns), nrows=len(dp)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n",
      "ERROR:    Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 50, in schema\n",
      "    id=pivot.obj.id, columns=_get_column_infos(dp, columns), nrows=len(dp)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n",
      "[2022-09-01 13:42:57,696][ERROR][uvicorn.error:406] :: Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 50, in schema\n",
      "    id=pivot.obj.id, columns=_get_column_infos(dp, columns), nrows=len(dp)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n",
      "ERROR:    Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 104, in rows\n",
      "    column_infos = _get_column_infos(dp, columns)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n",
      "[2022-09-01 13:42:57,718][ERROR][uvicorn.error:406] :: Exception in ASGI application\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py\", line 403, in run_asgi\n",
      "    result = await app(self.scope, self.receive, self.send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py\", line 78, in __call__\n",
      "    return await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/applications.py\", line 269, in __call__\n",
      "    await super().__call__(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/applications.py\", line 124, in __call__\n",
      "    await self.middleware_stack(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 184, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/errors.py\", line 162, in __call__\n",
      "    await self.app(scope, receive, _send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 92, in __call__\n",
      "    await self.simple_response(scope, receive, send, request_headers=headers)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/middleware/cors.py\", line 147, in simple_response\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 93, in __call__\n",
      "    raise exc\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/exceptions.py\", line 82, in __call__\n",
      "    await self.app(scope, receive, sender)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n",
      "    raise e\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 670, in __call__\n",
      "    await route.handle(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 266, in handle\n",
      "    await self.app(scope, receive, send)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/routing.py\", line 65, in app\n",
      "    response = await func(request)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 227, in app\n",
      "    raw_response = await run_endpoint_function(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n",
      "    return await run_in_threadpool(dependant.call, **values)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/starlette/concurrency.py\", line 41, in run_in_threadpool\n",
      "    return await anyio.to_thread.run_sync(func, *args)\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/to_thread.py\", line 31, in run_sync\n",
      "    return await get_asynclib().run_sync_in_worker_thread(\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n",
      "    return await future\n",
      "  File \"/Users/eyubogln/miniconda3/envs/domino/lib/python3.8/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n",
      "    result = context.run(func, *args)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 104, in rows\n",
      "    column_infos = _get_column_infos(dp, columns)\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 73, in _get_column_infos\n",
      "    return [\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/api/routers/datapanel.py\", line 78, in <listcomp>\n",
      "    cell_props=dp[col].formatter.cell_props,\n",
      "  File \"/Users/eyubogln/code/meerkat/meerkat/interactive/formatter.py\", line 82, in cell_props\n",
      "    \"dtype\": self.dtype,\n",
      "AttributeError: 'BasicFormatter' object has no attribute 'dtype'\n"
     ]
    }
   ],
   "source": [
    "interface.launch()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8e4263e10f94575b28c23f90f611f98",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "out = mk.embed(data=mk.PandasSeriesColumn([\"hello\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 512)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5:16:35 PM [vite] hmr update /src/lib/component/match/Match.svelte\n",
      "hmr update /src/routes/__layout.svelte\n",
      "5:17:16 PM [vite] hmr update /src/lib/component/match/Match.svelte\n",
      "hmr update /src/app.css\n",
      "5:17:22 PM [vite] hmr update /src/lib/component/match/Match.svelte\n",
      "hmr update /src/app.css\n",
      "5:19:25 PM [vite] hmr update /src/lib/component/match/Match.svelte\n",
      "hmr update /src/app.css\n"
     ]
    }
   ],
   "source": [
    "out.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('domino')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7dc70abeb586db4fbbb19fd8ec040a8328b2a3c884b1c8ab3f626b1589cb18a0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
